src/pool/srv_pool.c

```diff
@@ -4844,6 +4839,274 @@ ds_pool_svc_check_evict(uuid_t pool_uuid, d_rank_list_t *ranks,
+		D_GOTO(out, rc);
+
+rechoose:
+	ep.ep_grp = NULL; /* primary group */
+	rc = rsvc_client_choose(&client, &ep);
+	if (rc != 0) {
+		D_ERROR(DF_UUID ": cannot find pool service: " DF_RC "\n",
+			DP_UUID(uuid), DP_RC(rc));
+		goto out_client;
+	}
+
+realloc_resp:
+	rc = pool_req_create(info->dmi_ctx, &ep, POOL_RANKS_GET, &rpc);
+	if (rc != 0) {
+		D_ERROR(DF_UUID ": failed to create POOL_RANKS_GET rpc, "
+			DF_RC "\n", DP_UUID(uuid), DP_RC(rc));
+		D_GOTO(out_client, rc);
+	}
+
+	/* Allocate response buffer */
+	out_ranks = d_rank_list_realloc(out_ranks, resp_nranks);
+	if (out_ranks == NULL)
+		D_GOTO(out_rpc, rc = -DER_NOMEM);
+
+	in = crt_req_get(rpc);
+	uuid_copy(in->prgi_op.pi_uuid, uuid);
+	uuid_clear(in->prgi_op.pi_hdl);
+	in->prgi_nranks = resp_nranks;
+	rc = ranks_get_bulk_create(info->dmi_ctx, &in->prgi_ranks_bulk,
+				   out_ranks->rl_ranks, in->prgi_nranks);
+	if (rc != 0)
+		D_GOTO(out_resp_buf, rc);
+
+	D_DEBUG(DF_DSMS, DF_UUID ": send POOL_RANKS_GET to PS rank %u, "
+		"reply capacity %u\n", DP_UUID(uuid), ep.ep_rank, resp_nranks);
+
+	rc = dss_rpc_send(rpc);
+	out = crt_reply_get(rpc);
+	D_ASSERT(out != NULL);
+
+	rc = rsvc_client_complete_rpc(&client, &ep, rc,
+				      out->prgo_op.po_rc,
+				      &out->prgo_op.po_hint);
+	if (rc == RSVC_CLIENT_RECHOOSE) {
+		/* To simplify logic, destroy bulk hdl and buffer each time */
+		ranks_get_bulk_destroy(in->prgi_ranks_bulk);
+		d_rank_list_free(out_ranks);
+		crt_req_decref(rpc);
+		dss_sleep(1000 /* ms */);
+		D_GOTO(rechoose, rc);
+	}
+
+	rc = out->prgo_op.po_rc;
+	if (rc == -DER_TRUNC) {
+		/* out_ranks too small - realloc with server-provided nranks */
+		resp_nranks = out->prgo_nranks;
+		ranks_get_bulk_destroy(in->prgi_ranks_bulk);
+		d_rank_list_free(out_ranks);
+		crt_req_decref(rpc);
+		dss_sleep(1000 /* ms */);
+		D_GOTO(realloc_resp, rc);
+	} else if (rc != 0) {
+		D_ERROR(DF_UUID ": failed to get ranks, " DF_RC "\n",
+			DP_UUID(uuid), DP_RC(rc));
+	} else {
+		out_ranks->rl_nr = out->prgo_nranks;
+		*ranks = out_ranks;
+	}
+
+	ranks_get_bulk_destroy(in->prgi_ranks_bulk);
+out_resp_buf:
+	if (rc != 0)
+		d_rank_list_free(out_ranks);
+out_rpc:
+	crt_req_decref(rpc);
+out_client:
+	rsvc_client_fini(&client);
+out:
+	return rc;
+}
+
+/* CaRT RPC handler run in PS leader to return pool storage ranks
+ */
+void
+ds_pool_ranks_get_handler(crt_rpc_t *rpc)
+{
+	struct pool_ranks_get_in	*in = crt_req_get(rpc);
+	struct pool_ranks_get_out	*out = crt_reply_get(rpc);
+	uint32_t			 nranks = 0;
+	d_rank_list_t			out_ranks = {0};
+	struct pool_svc			*svc;
+	int				 rc;
+
+	D_DEBUG(DF_DSMS, DF_UUID ": processing rpc %p:\n",
+		DP_UUID(in->prgi_op.pi_uuid), rpc);
+
+	rc = pool_svc_lookup_leader(in->prgi_op.pi_uuid, &svc,
+				    &out->prgo_op.po_hint);
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+	/* This is a server to server RPC only */
+	if (daos_rpc_from_client(rpc))
+		D_GOTO(out, rc = -DER_INVAL);
+
+	rc = ds_pool_get_ranks(in->prgi_op.pi_uuid, MAP_RANKS_UP, &out_ranks);
```

Comment:
Agree with your concern (on Skype) that this is potentially racy (due to trying to hold pool->sp_lock). Even if we were to fix that by reading from the DB, by the time the reply reaches the MS rank, the pool map may still have changed. So this is an inherent disadvantage of our approach here.